yt/frontends/cholla/data_structures.py

```diff
@@ -9,16 +13,193 @@
+        # this branch covers distributed datasets (directly written by Cholla) and
+        # older concatenated files.
+        #
+        # historically, when the dataset is concatenated (in post-processing),
+        # the "nprocs" hdf5 attribute has been dropped
+        blockid_location_arr = _infer_blockid_location_arr(
+            fname_template=inferred_fname_template,
+            global_dims=f.attrs["dims"].astype("=i8"),
+            arr_shape=f.attrs.get("nprocs", np.array([1, 1, 1])).astype("=i8"),
+        )
+        consolidated_data = blockid_location_arr.size == 1
+
+        def _get_common_idx():
+            return (slice(None), slice(None), slice(None))
+
+        field_idx_map = defaultdict(_get_common_idx)
+
+    # STEP 4: Finalize the fname template
+    # ===================================
+    if consolidated_data:
+        fname_template = filename
+    elif cur_filename_suffix != 0:
+        raise ValueError(  # mostly just a sanity check!
+            "filename passed to yt.load for a distributed cholla dataset must "
+            "end in '.0'"
+        )
+    else:
+        fname_template = inferred_fname_template
+
+    mapping = _BlockDiskMapping(
+        fname_template=fname_template,
+        field_group="" if flat_structure else "field",
+        field_idx_map=field_idx_map,
+    )
+    return blockid_location_arr, mapping
+
+
+def _split_fname_procid_suffix(filename: str):
+    """Splits ``filename`` at the '.' separating the beginning part of the
+    string from the process-id suffix, and returns both parts in a 2-tuple.
+
+    There are 2 conventions for the names of Cholla's data-files:
+
+    1. "root.h5.{procid}" is the standard format used when Cholla directly
+       writes out data-files. When outputting a snaphsot, each MPI-process
+       writes a separate file, named using this template (``{procid}`` is
+       replaced with MPI-rank). Modern versions of Cholla without MPI replace
+       ``{procid}`` with ``0``. When passed a name of this form, the function
+       returns ``("root.h5", "{procid}")``
+    2. "root.h5": is the standard format used by Cholla's concatenation scirpts
+       (older versions of the code compiled without MPI also used this format
+       to name outputs). When passed a name of this form, the function returns
+       ``("root.h5", "")``
+
+    Notes
+    -----
+    The following examples illustrate how the behavior differs from that of
+    ``os.path.splitext``
+
+    - For "root.h5.3": ``os.path.splitext`` returns ``("root.h5", ".3")``,
+      while this function returns ``("root.h5", "3")``
+    - For "root.h5": os.path.splitext`` returns ``("root", ".h5")``, while this
+      function returns ``("root.h5", "")``
+    """
+
+    # at this time, we expect the suffix to be the minimum number of characters
+    # that are necessary to represent the process id. For flexibility, we will
+    # allow extra zero-padding
+
+    sep_i = filename.rfind(".")
```

Comment:
I think this function could be much more readable if it was written around `str.rsplit` instead of `str.rfind`

Reply:
or better yet: `str.rpartition`